client/extensions/woocommerce/app/store-stats/referrers/index.js

```diff
@@ -21,61 +21,165 @@ import StoreStatsPeriodNav from 'woocommerce/app/store-stats/store-stats-period-
-			<StoreStatsPeriodNav
-				type="referrers"
-				selectedDate={ selectedDate }
-				unit={ unit }
-				slug={ slug }
-				query={ query }
-				statType={ STAT_TYPE }
-				title={ translate( 'Store Referrers' ) }
-				queryParams={ queryParams }
-			/>
-			<Module
-				siteId={ siteId }
-				emptyMessage={ translate( 'No referrers found' ) }
-				query={ query }
-				statType={ STAT_TYPE }
-			>
-				<table>
-					<tbody>
-						{ sortedData.map( d => (
-							<tr key={ d.referrer }>
-								<td>{ d.date }</td>
-								<td>{ d.referrer }</td>
-								<td>{ d.product_views }</td>
-								<td>{ d.add_to_carts }</td>
-								<td>{ d.product_purchases }</td>
-								<td>${ d.sales }</td>
-							</tr>
-						) ) }
-					</tbody>
-				</table>
-			</Module>
-			<JetpackColophon />
-		</Main>
-	);
-};
+class Referrers extends Component {
+	static propTypes = {
+		siteId: PropTypes.number,
+		query: PropTypes.object.isRequired,
+		data: PropTypes.array.isRequired,
+		selectedDate: PropTypes.string,
+		unit: PropTypes.oneOf( [ 'day', 'week', 'month', 'year' ] ),
+		slug: PropTypes.string,
+		limit: PropTypes.number,
+	};
+
+	state = {
+		filter: '',
+	};
+
+	onSearch = str => {
+		this.setData( this.props, str );
+	};
+
+	getFilteredSortedData = ( filter, { data, selectedDate, unit } ) => {
+		const unitSelectedDate = getUnitPeriod( selectedDate, unit );
+		const selectedData = find( data, d => d.date === unitSelectedDate ) || { data: [] };
+		const filteredData = filter
+			? selectedData.data.filter( d => d.referrer.match( filter ) )
+			: selectedData.data;
+		return {
+			filteredSortedData: sortBySales( filteredData ),
+			unfilteredDataLength: selectedData.data.length,
+		};
+	};
+
+	getSelectedReferrer = ( filteredSortedData, { queryParams } ) => {
+		let selectedReferrerIndex = null;
+		const selectedReferrer = find( filteredSortedData, ( d, idx ) => {
+			if ( queryParams.referrer && queryParams.referrer === d.referrer ) {
+				selectedReferrerIndex = idx;
+				return true;
+			}
+			return false;
+		} );
+		return {
+			selectedReferrer,
+			selectedReferrerIndex,
+		};
+	};
+
+	setData( props, _filter ) {
+		const isEmpty = _filter === '';
+		const filter = isEmpty ? '' : _filter || this.state.filter;
+		const { filteredSortedData, unfilteredDataLength } = this.getFilteredSortedData(
+			filter,
+			props
+		);
+		const { selectedReferrer, selectedReferrerIndex } = this.getSelectedReferrer(
+			filteredSortedData,
+			props
+		);
+		this.setState( {
+			filter,
+			filteredSortedData,
+			unfilteredDataLength,
+			selectedReferrer,
+			selectedReferrerIndex,
+		} );
+	}
 
-Referrers.propTypes = {
-	siteId: PropTypes.number,
-	query: PropTypes.object.isRequired,
-	data: PropTypes.array.isRequired,
-	selectedDate: PropTypes.string,
-	unit: PropTypes.oneOf( [ 'day', 'week', 'month', 'year' ] ),
-	slug: PropTypes.string,
-};
+	componentWillReceiveProps( nextProps ) {
```

Comment:
Random preference - but I like having lifecycle methods declared near the top of the component. Not a written standard but seems to be the common practice in calypso.